### ISSUE TITLE
Possible issue in zonbudusg.exe

### ISSUE THREAD

**Issue by hjia1005** [MODFLOW-ORG/mfusg#50]
This issue is found during a discussion in Flopy:
[https://github.com/modflowpy/flopy/discussions/1447](url)

I was doing the zone budget for my mfusg model and found inconsistent results when using zonbudusg.exe, class CellBudgetFile in FloPy and usgbud2smp.exe from PEST gw utility programs, while the latter two had the same results. I literally applied 100% of rainfall over some mine pits (different zones in my model) but some of them got very low or even 0 recharge from the zonbudusg.exe calculation. The inconsistency only happens to ITYPE==3 packages. After a long discussion with @cnicol-gwlogic, we feel that there may be a bug in the source code of zonbudusg.exe. I pasted the code below.

![image](https://user-images.githubusercontent.com/33449394/180599907-afdac988-a15b-45a4-90cb-a23d8926a205.png)

I think the cause of the inconsistency could be with the repeated use of array BUFFD. It is very possible that BUFFD(I)=DZERO reset some of the previously assigned rch/evt to 0. Another temporary array variable should have been used to read the ITYPE==3 arrays and then pass it to BUFFD. Then I manually checked the zones in my model. If most or all of the gnode numbers in a specific zone are smaller than the size of rch arrays (66120 in my case, also the number of rnodes) then the accumulated recharge or evaporation is very small or 0. Zone2 to zone4 in my model are like this and zone 4 got 0. If most or all of the gnode numbers in a specific zone are larger than the size of rch arrays (66120 in my case) they are well preserved because there is no chance for them to be re-assigned to 0 by BUFFD(I)=DZERO, such as zone 5 to zone 7 in my model. 

@langevin-usgs suggested me to create a simple model to report the issue. I modified the mfusg example in flopy3_gridgen.ipynb by removing some cells in the top layer and created the gnode-rnode connection. Then I applied RCH which covers the top cells of the entire model - exactly the same approach used in my real model. However this time the result from zonbudusg.exe is consistent with the result from CellBudgetFile in Flopy. I defined many zonations (zone by layer, or random zone numbers) to test it and they all came back right. Then I exported the cbb recharge arrays to csv and manually simulated how zonbudusg.exe should be processing the data based on the source codes. This can be done manually only because this simple model has 485 gnodes and 117 rnodes. Then I realize that in this simple model, after BUFFD(I) is set to 0, they always have a chance to be reset to the real value later by BUFFD(IBUFF(I))=BUFFD(I), it's like in this simple model, gnodes and rnodes are playing peacefully. But in the real model with 401091 gnodes and 66120 rnodes, there are many conflicts between them because the same array variable BUFFD is used. My real model files are very big and they are project related, not sure if it's proper to send them over.

Anyway, this is just based on my understanding of the source code in Fortran after I got this issue. We spent quite a long time testing and discussing it. I wish the developers may look into it but I am not able to provide an example for now.

**Comment by christianlangevin** [comment]
Do you have a flopy script that will generate a model that demonstrates the issue?  If not, can you create a very small test problem and zip and upload both the mfusg and zonbudusg input files?  It would be greatly appreciated.

**Comment by hjia1005** [comment]
@langevin-usgs I tried very hard to modify the simple model to many versions, but zonbudusg.exe seems to work fine with a simple model with only a few hundred nodes because the conflicts between gnode and rnode can always be avoided due to less number of nodes. Therefore I submitted my own model files for you to identify the issue. The zip file is over 25MB so have to upload it to my own repository.

[https://github.com/hjia1005/zonbudusg_test_model](https://github.com/hjia1005/zonbudusg_test_model)

The model.disu file was modified to have only one stress period to reduce the run time, but still needs 5 to 7 mins to run. model.cbb was not included because of the size is over 1GB). I include the model files and the zonbudusg input and result files.

- model list file - model.lst
- zone file - zbud.zon
- zonbudusg input file - zbud.rsp
- zonbudusg list file - zbud.list

To make it simple to compare, zbud.zon only has a single zone, zone number = 1 is assigned to all model cells.

model.lst shows that the recharge is 5288.9780 m3/day.
![image](https://user-images.githubusercontent.com/33449394/180633250-e51f0cc9-a4c9-4959-a68e-a357563a5e07.png)

zbud.list shows that the recharge is 4274.4 m3/day, which should be the same as model.lst since there is only one zone.
![image](https://user-images.githubusercontent.com/33449394/180633211-e3fa4a5b-3aa1-4ee4-a8c8-af866a428a1e.png)

I wish this is enough to clarify the issue I had. Thanks for your time.
